### ISSUE TITLE
Metadata not counted as evictable/not being reclaimed. Leaked?

### ISSUE THREAD

**Issue by dbavatar** [openzfs/zfs#6506]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | x
Distribution Version    | x
Linux Kernel                 | 4.4 through 4.9.24+
Architecture                 | x86_64
ZFS Version                  | 0.7.1
SPL Version                  | 0.7.1

### Describe the problem you're observing
Metadata is not being counted as evictable, and is not evicting. In fact, I can't get it to reduce in size short of deleting the files or unmounting, i.e. pressure from arc data has no effect. As a side note, 7.x seems to also be using twice as much memory per file as 6.5.x, although that's not part of this bug.

I've bisected this back to commit d3c2ae1c0806b183a315e3d43cc8018cfdca79b5. Unfortunately it's 4500 lines.

### Describe how to reproduce the problem
1. Load zfs cleanly and import non-root zpool such that arcstats are minimal.
2. Create enough files to measure effect with high confidence, in my case 3 dirs with 32k files each, 0 bytes, uses ~64MB to ~128MB  To refill metadata after reload: "ls -ltr [path]* > /dev/null"
3. Observe that arcstats shows a large amount of size/metadata_size and mru_size/mfu_size. However mru_evictable* and mfu_evictable* are tiny.
4. Observe that echo 3 > /proc/sys/vm/drop_caches does not shrink metadata_size.

My bisect script:

```sh
sh autogen.sh && ./configure --with-config=kernel  --prefix=/ && make -j 10 && make install
zpool export -a
killall zed; rmmod zfs zunicode zcommon zavl znvpair icp spl;
depmod -a
modprobe zfs
zpool import -a
ls -ltr /tank/d* > /dev/null
echo 3 > /proc/sys/vm/drop_caches
sleep 1
echo 3 > /proc/sys/vm/drop_caches
sleep 1
echo 3 > /proc/sys/vm/drop_caches
cat /proc/spl/kstat/zfs/arcstats |grep ^metadata_size | awk '{ if ($3 > 50000000) { print "Bad: " $3; exit 1} print "Good: " $3 }'
```

**Comment by dbavatar** [comment]
@grwilson for d3c2ae1c0806b183a315e3d43cc8018cfdca79b5

**Comment by dbavatar** [comment]
@behlendorf I think this will have to go in a stable release. I've been looking at this all day, looks pretty bad. Operates effectively as a memory leak, metadata for every file that exists stays in memory, growing past arc limits.

**Comment by behlendorf** [comment]
@dbavatar if that's the case, we should definitely get a fix for this in to the next point release.  But first let's determine exactly what is going wrong and take it from there.

It's hard to imagine meta-data is never getting reclaimed or everything would grind to a halt pretty quickly.  Likely something a little more subtle is going on.  Since you identified the compressed ARC change as where the regression was introduced it would be useful to determine if disabling the compressed arc by setting the module option `zfs_compressed_arc_enabled=0` changes the behavior.

@dweeezil might be interested in this issue as well.

**Comment by zviratko** [comment]
would this only affect filesystems, or would it include volumes as well? I have tens of GBs of memory hoarded somewhere in kernel slab and drop_caches doesn't free that neither. My pool has 0 files, but hundreds of ZVOLs and ~20K snaps. arcstats show lots of metadata IO (served from cache and linux cache), so I guess any metadata-heavy workload would be affected?

**Comment by bunder2015** [comment]
ZVOLs also use ARC unless you set `primarycache=none`

**Comment by zviratko** [comment]
I know they do but I was refering to the reported memory leak - could it be the same issue or would this only apply when ZFS is used as a filesystem? What I see is that there are maybe 30-40G tied somewhere in kernel that don't get freed, but it doesn't seem to be rising much (I can go few months without a reboot on a 128G machine that also has 64G ARC).

**Comment by dweeezil** [comment]
I'm going to try to reproduce this.  @dbavatar Could you please also grab ```dbuf_size``` through ```anon_size``` from the arcstats as well.

**Comment by chrisrd** [comment]
@dbavatar I'm not sure if I'm able to replicate your problem w/ linux 4.9.43 + ZFS today's HEAD 

tl;dr: `metadata_size` is `2,640,896` on fresh module load, goes up to `72,632,832` after reading the large directories, and drops down to `17,822,208` a few seconds after `drop_caches`. Note that it *CAN* take more than a second as part of the freeing-up memory process is asynchronous.

Is `metadata_size` of `17k` after `drop_caches` showing your problem, or is it showing your problem isn't occurring here?

VM with 1G RAM, `linux-4.9.43` and ZFS `0.7.0-31_gc8f9061`

Pool and directories created using:
```
f=/var/tmp/file-vdev0
truncate -s 4G "${f}"
loop=$(losetup -f)
losetup "${loop}" "${f}"
zpool create testpool "${loop}"
zfs create testpool/foo
for i in {0..2}; do
  mkdir /testpool/foo/$i
  for j in {0..32768}; do
    touch /testpool/foo/$i/$j
  done
done
```

`arcstats` values extracted using:
```
awk '$1~/(^|_)size|evictable|arc_meta/ {print $1, $3}' /proc/spl/kstat/zfs/arcstats | sort
```

Results (using a little `perl` and `paste`):
```
col 1: field
col 2: after fresh module unload/load
col 3: after 'find /testpool/foo -mtime +100'
col 4: after drop_caches
col 5: after a further 10m (little to no activity)
col 6: after a further 10m (little to no activity) + drop_caches

anon_size                         32,768       32,768       32,768       32,768       32,768
arc_meta_limit               387,930,624  387,930,624  387,930,624  387,930,624  387,930,624
arc_meta_max                   3,013,248  277,181,800  331,429,544  331,429,544  331,429,544
arc_meta_min                  16,777,216   16,777,216   16,777,216   16,777,216   16,777,216
arc_meta_used                  2,941,784  331,429,544   94,424,424   94,424,424   94,023,096
bonus_size                        15,680   31,474,880    9,110,720    9,110,720    9,080,000
compressed_size                  203,776    6,449,664      741,888      741,888      638,976
dbuf_size                         61,304   64,627,056   18,572,584   18,572,584   18,510,016
dnode_size                       171,296  159,040,256   46,099,632   46,099,632   45,944,496
hdr_size                          52,608    3,654,520    2,819,280    2,819,280    2,818,440
metadata_size                  2,640,896   72,632,832   17,822,208   17,822,208   17,670,144
mfu_evictable_metadata                 0            0       76,288       76,288        1,536
mfu_ghost_evictable_metadata           0            0   22,790,144   22,790,144   25,214,976
mfu_ghost_size                         0            0   22,790,144   22,790,144   25,214,976
mfu_size                         745,472   36,303,360   10,408,960   10,408,960   10,283,520
mru_evictable_metadata            90,112        1,024       26,624       26,624            0
mru_ghost_evictable_metadata           0            0   23,070,208   23,070,208   23,922,176
mru_ghost_size                         0            0   23,070,208   23,070,208   23,922,176
mru_size                       1,862,656   36,296,704    7,380,480    7,380,480    7,353,856
overhead_size                  2,437,120   66,183,168   17,080,320   17,080,320   17,031,168
size                           2,941,784  331,429,544   94,424,424   94,424,424   94,023,096
uncompressed_size              2,766,336   66,217,472   20,357,120   20,357,120   17,080,320
```

**Comment by ssergiienko** [comment]
Maybe data from my real working system could help.
Initial state:
```bash
awk '$1~/(^|_)size|evictable|arc_meta/ {print $1, $3}' /proc/spl/kstat/zfs/arcstats | sort
anon_evictable_data 0
anon_evictable_metadata 0
anon_size 3757056
arc_meta_limit 6265697280
arc_meta_max 625879240
arc_meta_min 16777216
arc_meta_used 489386296
bonus_size 50374400
compressed_size 1728507392
data_size 1762442752
dbuf_size 51127936
dnode_size 143481728
hdr_size 20227128
l2_hdr_size 0
l2_size 0
metadata_size 224175104
mfu_evictable_data 522652672
mfu_evictable_metadata 15692800
mfu_ghost_evictable_data 0
mfu_ghost_evictable_metadata 0
mfu_ghost_size 0
mfu_size 623808512
mru_evictable_data 1113472000
mru_evictable_metadata 10160640
mru_ghost_evictable_data 0
mru_ghost_evictable_metadata 0
mru_ghost_size 0
mru_size 1359052288
overhead_size 258110464
size 2251829048
uncompressed_size 3350349824
```
Just after `drop_caches 3`:
```bash
awk '$1~/(^|_)size|evictable|arc_meta/ {print $1, $3}' /proc/spl/kstat/zfs/arcstats | sort
anon_evictable_data 0
anon_evictable_metadata 0
anon_size 2530304
arc_meta_limit 6265697280
arc_meta_max 625879240
arc_meta_min 16777216
arc_meta_used 287058896
bonus_size 30048640
compressed_size 23734272
data_size 17670656
dbuf_size 30420672
dnode_size 88289344
hdr_size 5555536
l2_hdr_size 0
l2_size 0
metadata_size 132744704
mfu_evictable_data 0
mfu_evictable_metadata 524288
mfu_ghost_evictable_data 0
mfu_ghost_evictable_metadata 124175360
mfu_ghost_size 124175360
mfu_size 24578048
mru_evictable_data 0
mru_evictable_metadata 2296320
mru_ghost_evictable_data 14150656
mru_ghost_evictable_metadata 203173376
mru_ghost_size 217324032
mru_size 123307008
overhead_size 126681088
size 304729552
uncompressed_size 14543616
```
Few seconds later matadata_size dropped even more - `metadata_size 92065280`

ZFS-git ~= 0.7.1
`uname -r
4.12.6-1-ARCH`

**Comment by dbavatar** [comment]
@behlendorf zfs_compressed_arc_enabled=0 is the first thing I tried, didn't help.
@ssergiienko @chrisrd size should have reduced to arc_c_min, with metadata_size being a subset of that on drop_caches, so unless that system has active IO, you have the problem.

@chrisrd I disagree that it should take more time to clear up more memory, drop_caches should be synchronous. Internally, it MUST be synchronous to avoid OOM problems in the kernel. This used to be problematic but I fixed this in 7.0. We're probably missing a basis for comparison here because I have the same fixes applied to my local 6.5.x branches. So you should apply drop_caches multiple times with sleep if looking at 6.5.x. However, this is not the issue in 7.1

I moved to a fresh system with one vdev where I keep the same kernel (4.4.70) and only switch between 6.5.8 and 7.1 ZFS, and I clearly see the issue. Although it seems to be better than the multiple vdev case for some reason.


6.5.8 + https://github.com/dbavatar/zfs/tree/dbavatar/cache_reclaim
```
c_min                           4    33554432
size                            4    134205776
hdr_size                        4    948400
data_size                       4    0
metadata_size                   4    35450880
other_size                      4    97806496
anon_size                       4    16384
anon_evictable_data             4    0
anon_evictable_metadata         4    0
mru_size                        4    28700672
mru_evictable_data              4    0
mru_evictable_metadata          4    28090368
```
after drop caches:
```
c_min                           4    33554432
size                            4    1024424
hdr_size                        4    758496
data_size                       4    0
metadata_size                   4    167936
other_size                      4    97992
anon_size                       4    16384
anon_evictable_data             4    0
anon_evictable_metadata         4    0
mru_size                        4    151552
mru_evictable_data              4    0
mru_evictable_metadata          4    0

```

7.1
```
c_min                           4    33554432
size                            4    88896784
compressed_size                 4    32560640
uncompressed_size               4    32560640
overhead_size                   4    15599616
hdr_size                        4    752592
data_size                       4    0
metadata_size                   4    48160256
dbuf_size                       4    8657376
dnode_size                      4    22751840
bonus_size                      4    8574720
anon_size                       4    32768
anon_evictable_data             4    0
anon_evictable_metadata         4    0
mru_size                        4    47501312
mru_evictable_data              4    0
mru_evictable_metadata          4    16826368
```
drop_caches:
```
c_min                           4    33554432
size                            4    71937808
compressed_size                 4    15601664
uncompressed_size               4    15601664
overhead_size                   4    15599616
hdr_size                        4    752592
data_size                       4    0
metadata_size                   4    31201280
dbuf_size                       4    8657376
dnode_size                      4    22751840
bonus_size                      4    8574720
anon_size                       4    32768
anon_evictable_data             4    0
anon_evictable_metadata         4    0
mru_size                        4    30674944
mru_evictable_data              4    0
mru_evictable_metadata          4    0

```

What to note here mru_evictable_metadata ~= mru_size in 6.5.8, and is completely cleaned up by drop caches. No longer true in 7.1, it should be shrinking down to the memory footprint of just after zpool import.

**Comment by dbavatar** [comment]
This is where it gets very concerning: 101*32768 = 3.3 million 0-byte files scanned via `ls -ltr`, leads to ~1.6GB of  arc "size". After drop_caches multiple times, we still have ~450MB reported arc "size", but ~720MB tied up by the pool. My biggest concern is not even with the memory footprint, but kernel allocation failures from not being able to shrink to arc_c_min on-demand.

```
cat /proc/spl/kstat/zfs/arcstats  |grep "size\|evictable\|min\|max"
hash_elements_max               4    114026
hash_chain_max                  4    2
c_min                           4    33554432
c_max                           4    8388417536
size                            4    1694289528
compressed_size                 4    1297934848
uncompressed_size               4    1297934848
overhead_size                   4    96536576
hdr_size                        4    27454280
data_size                       4    0
metadata_size                   4    1394471424
dbuf_size                       4    58823232
dnode_size                      4    155047472
bonus_size                      4    58493120
anon_size                       4    32768
anon_evictable_data             4    0
anon_evictable_metadata         4    0
mru_size                        4    1389072384
mru_evictable_data              4    0
mru_evictable_metadata          4    1199865856
mru_ghost_size                  4    28360704
mru_ghost_evictable_data        4    0
mru_ghost_evictable_metadata    4    28360704
mfu_size                        4    5366272
mfu_evictable_data              4    0
mfu_evictable_metadata          4    1530368
mfu_ghost_size                  4    0
mfu_ghost_evictable_data        4    0
mfu_ghost_evictable_metadata    4    0
l2_size                         4    0
l2_asize                        4    0
l2_hdr_size                     4    0
arc_meta_max                    4    7036048152
arc_meta_min                    4    16777216

```
After multiple ```echo 3 > /proc/sys/vm/drop_caches```
```
cat /proc/spl/kstat/zfs/arcstats  |grep "size\|evictable\|min\|max"
hash_elements_max               4    114026
hash_chain_max                  4    2
c_min                           4    33554432
c_max                           4    8388417536
size                            4    473593496
compressed_size                 4    96538624
uncompressed_size               4    96538624
overhead_size                   4    96536576
hdr_size                        4    8249192
data_size                       4    0
metadata_size                   4    193075200
dbuf_size                       4    58803264
dnode_size                      4    154993200
bonus_size                      4    58472640
anon_size                       4    32768
anon_evictable_data             4    0
anon_evictable_metadata         4    0
mru_size                        4    189140992
mru_evictable_data              4    0
mru_evictable_metadata          4    0
mru_ghost_size                  4    283983872
mru_ghost_evictable_data        4    0
mru_ghost_evictable_metadata    4    283983872
mfu_size                        4    3901440
mfu_evictable_data              4    0
mfu_evictable_metadata          4    0
mfu_ghost_size                  4    1497600
mfu_ghost_evictable_data        4    0
mfu_ghost_evictable_metadata    4    1497600
l2_size                         4    0
l2_asize                        4    0
l2_hdr_size                     4    0
arc_meta_max                    4    7036048152
arc_meta_min                    4    16777216
```

How much memory is the pool actually using now? Look at delta used.
```
$ free
             total       used       free     shared    buffers     cached
Mem:      16383628    1290216   15093412        388       4372     110332
-/+ buffers/cache:    1175512   15208116
Swap:      4194300          0    4194300

$ zpool export -a

$ free
             total       used       free     shared    buffers     cached
Mem:      16383628     553720   15829908        388       4512     111936
-/+ buffers/cache:     437272   15946356
Swap:      4194300          0    4194300
```

I don't see dbuf_size of dnode_size move at all, do we need reclaim implementations for those?

**Comment by sempervictus** [comment]
This is the sort of thing that hurts reputation if not addressed post-haste in releases. Large NFS setups used by hospitals will get murdered by this. Buildbots need strict mm testing most likely - the 0 size files taking up space after evict seems a decent way to verify such things.

**Comment by dweeezil** [comment]
@dbavatar I'm finally in a position to start looking into this:  do you know whether the system is using memory cgroups, and if so, whether the issue occurs if they are disabled?  I remember while working on the problem which was fixed by 09fb30e5e9 that it can be very difficult to free memory when memcg's are being used.  Also, although you've only been describing the ARC state following a ```drop_slab()``` operation, do I gather correctly that the actual issue is that other more "normal" calls to the shrinker are not freeing memory that ought to be freeable?

Unfortunately, I'm not really very familiar memory cgroups:  the issue involving the commit I referenced above was one of the first times I seriously looked into them.  It's also possible this problem has nothing to do with memcg's.  In the mean time, I'm going to get some tests set up to reproduce this.

**Comment by dbavatar** [comment]
@dweeezil I have memory cgroups compiled in, but disabled. I did test with it (coincidentally) enabled, it made no difference.

**Comment by dweeezil** [comment]
@dbavatar Thanks for the information.  I'll get a new round of tests run tomorrow after disabling most debugging.  I've been configuring my test system with ```--enable-debug --enable-debug-kmem --enable-debug-kmem-tracking``` which makes this type of testing unbearably slow.

**Comment by chrisrd** [comment]
I can confirm the bisect by @dbavatar : `metadata_size` definitely drops a lot more before commit d3c2ae1 than after. Using same test setup as before, and on linux-4.4:

```
col 1: field
col 2: after fresh module unload/load
col 3: after 'find /testpool/foo -mtime +100'
col 4: after drop_caches


# before, @b8eb3c4e
arc_meta_used             921,720  316,997,768    2,117,056
metadata_size             694,272   59,348,992      366,592

# after, @d3c2ae1c
arc_meta_used             875,672  329,340,872   91,692,192
metadata_size             651,264   70,572,032   15,744,000
```

`dbufstat.py` after each test may be instructive:

```
# before, @b8eb3c4e
$ dbufstat -b | sort -n -k1,5
           pool    objset      object  level     blkid        offset   dbsize
       testpool         0           0      0         0             0      16K
       testpool         0           0      0         1           16K      16K
       testpool         0           0      0         2           32K      16K
       testpool         0          61      0         0             0      16K
       testpool        51           0      0         0             0      16K
       testpool        51           0      1         0             0      16K
       testpool        51           0      2         0             0      16K
       testpool        51           0      3         0             0      16K
       testpool        51           0      4         0             0      16K
       testpool        51           0      5         0             0      16K
       testpool        51           0      6         0             0      16K
       testpool        74           0      0         0             0      16K
       testpool        74           0      0       584          9.0M      16K
       testpool        74           0      0      1710           26M      16K
       testpool        74           0      1         0             0      16K
       testpool        74           0      1         4           64K      16K
       testpool        74           0      1        13          208K      16K
       testpool        74           0      2         0             0      16K
       testpool        74           0      3         0             0      16K
       testpool        74           0      4         0             0      16K
       testpool        74           0      5         0             0      16K
       testpool        74           0      6         0             0      16K

# after, @d3c2ae1c
$ dbufstat -b | sort -n -k1,5
           pool    objset      object  level     blkid        offset   dbsize
       testpool         0           0      0         0             0      16K
       testpool         0           0      0         1           16K      16K
       testpool         0           0      0         2           32K      16K
       testpool         0           1      0         0             0      16K
       testpool         0           1      0         1           16K      16K
       testpool         0          61      0         0             0      16K
       testpool        51           0      0         0             0      16K
       testpool        51           0      1         0             0      16K
       testpool        51           0      2         0             0      16K
       testpool        51           0      3         0             0      16K
       testpool        51           0      4         0             0      16K
       testpool        51           0      5         0             0      16K
       testpool        51           0      6         0             0      16K
       testpool        74           0      0         0             0      16K
       testpool        74           0      0       584          9.0M      16K
       testpool        74           0      0      1024           16M      16K
       testpool        74           0      0      1025           16M      16K
       testpool        74           0      0      1026           16M      16K
       testpool        74           0      0      1027           16M      16K
       testpool        74           0      0      1028           16M      16K
...
       testpool        74           0      0      2045           31M      16K
       testpool        74           0      0      2046           31M      16K
       testpool        74           0      0      2048           32M      16K
       testpool        74           0      1         0             0      16K
       testpool        74           0      1         4           64K      16K
       testpool        74           0      1         8          128K      16K
       testpool        74           0      1         9          144K      16K
       testpool        74           0      1        10          160K      16K
       testpool        74           0      1        11          176K      16K
       testpool        74           0      1        12          192K      16K
       testpool        74           0      1        13          208K      16K
       testpool        74           0      1        14          224K      16K
       testpool        74           0      1        15          240K      16K
       testpool        74           0      1        16          256K      16K
       testpool        74           0      1        17          272K      16K
       testpool        74           0      1        18          288K      16K
       testpool        74           0      1        19          304K      16K
       testpool        74           0      1        20          320K      16K
       testpool        74           0      1        21          336K      16K
       testpool        74           0      1        22          352K      16K
       testpool        74           0      1        23          368K      16K
       testpool        74           0      2         0             0      16K
       testpool        74           0      3         0             0      16K
       testpool        74           0      4         0             0      16K
       testpool        74           0      5         0             0      16K
       testpool        74           0      6         0             0      16K
```
...where the `...` was another 870-odd lines at the same objset / object / level but increasing blkids (with gaps in the sequence).

**Comment by chrisrd** [comment]
d3c2ae1 added some `arcstat` fields: `compressed_size`, `overhead_size` and `uncompressed_size`. The "after" test shows `overhead_size` is uncomfortably close to the blown-out `metadata_size`:

```
col 1: field
col 2: after fresh module unload/load
col 3: after 'find /testpool/foo -mtime +100'
col 4: after drop_caches

# after, @d3c2ae1c
arc_meta_used             875,672  329,340,872   91,692,192
metadata_size             651,264   70,572,032   15,744,000
compressed_size            75,264    6,430,208      713,728   
overhead_size             576,000   64,141,824   15,030,272
uncompressed_size         675,840   64,176,128   18,421,760
```
In fact, the final `compressed_size + overhead_size` precisely matches  the `metadata_size` - I'm guessing that's no coincidence.

I've no idea yet what this all means!

**Comment by loli10K** [comment]
With https://github.com/zfsonlinux/zfs/commit/d3c2ae1c0806b183a315e3d43cc8018cfdca79b5 the old `arc_user_evicts_thread`/`arc_do_user_evicts()`/`arc_buf_eviction_needed()` are gone, replaced by `dbuf_evict_thread` which has a different eviction policy.

Also `arc_buf_t { .b_efunc }` (`arc_evict_func_t` callback function) is gone, i think this was the old way to trigger dbuf eviction from the ARC when we're under memory pressure.

I'm working on a patch to change the dbuf eviction policy when the ARC shrinks: this might not be the right way to fix this, i have literally a couple of days of experience on this code.

**Comment by chrisrd** [comment]
d3c2ae1 added a "[dbuf cache](https://github.com/zfsonlinux/zfs/blob/d3c2ae1c0806b183a315e3d43cc8018cfdca79b5/module/zfs/dbuf.c#L101)" with associated zfs parameters (undocumented in `man5/zfs-module-parameters.5` - that's a bit naughty!) including `dbuf_cache_max_bytes` and `dbuf_cache_max_shift`. On my 1GB test box, these default to:

```
# grep -H . /sys/module/zfs/parameters/{dbuf_cache_max_bytes,dbuf_cache_max_shift}
/sys/module/zfs/parameters/dbuf_cache_max_bytes:16162048
/sys/module/zfs/parameters/dbuf_cache_max_shift:5
```

After running the same `find` + `drop_caches`  tests as above, but using different `dbuf_cache_max_bytes` on module load, we get:
```
col 2: dbuf_cache_max_bytes=16162048 (default, 15.4M)
col 3: dbuf_cache_max_bytes=5242880 (5M)

anon_size                          32,768       16,384
arc_meta_limit                387,889,152  387,889,152
arc_meta_max                  329,340,872  318,356,056
arc_meta_min                   16,777,216   16,777,216
arc_meta_used                  91,692,192   33,544,696  <<<
bonus_size                      9,029,440    2,988,800
compressed_size                   713,728    1,364,480
dbuf_size                      18,412,688    6,103,224  <<<
dnode_size                     45,673,008   15,242,864
hdr_size                        2,833,056    2,661,840
metadata_size                  15,744,000    6,547,968  <<<
mfu_evictable_metadata             89,088      814,080
mfu_ghost_evictable_metadata   22,839,296    8,896,512
mfu_ghost_size                 22,839,296    8,896,512
mfu_size                       12,575,232    4,807,680
mru_evictable_metadata             37,376      313,856
mru_ghost_evictable_metadata   22,915,072   16,285,696
mru_ghost_size                 22,915,072   16,285,696
mru_size                        3,136,000    1,723,904
overhead_size                  15,030,272    5,183,488
size                           91,692,192   33,544,696
uncompressed_size              18,421,760   39,000,064
```

It looks like there's no mechanism to drop the dbuf cache below it's [low water mark](https://github.com/zfsonlinux/zfs/blob/d3c2ae1c0806b183a315e3d43cc8018cfdca79b5/module/zfs/dbuf.c#L128), e.g. in response to `echo 3 > /proc/sys/vm/drop_caches`. It's not clear if there should be or not.

I.e. it looks like the increase in `metadata_size` seen after d3c2ae1 is deliberate - mostly if not entirely due to the dbuf cache - rather than a leak per the topic of this ticket.

Resolved?

**Comment by tuxoko** [comment]
I kind of glance through the changes in d3c2ae1, it seems they added a lru list for dbuf, and it has a size of 100M. Those 100M will pin ARC down. I wonder if we should make this value float according to arc size, so that when we do shrink, we would shrink dbuf list as well...

**Comment by chrisrd** [comment]
Like this? https://github.com/chrisrd/zfs/commit/9fa593d

```
col 2: after fresh module unload/load
col 3: after 'find /testpool/foo -mtime +100'
col 4: after drop_caches

a: pre d3c2ae1
b: post d3c2ae1
c: Use arc_c instead of arc_c_max to scale dbuf cache
----

a: arc_meta_used                       921,720  316,997,768    2,117,056
b: arc_meta_used                       875,672  329,340,872   91,692,192
c: arc_meta_used                       855,744  330,057,864   26,355,696

a: metadata_size                       694,272   59,348,992      366,592
b: metadata_size                       651,264   70,572,032   15,744,000
c: metadata_size                       626,688   70,533,632    4,937,216

a: dbuf_size                            43,736   62,809,824       88,088
b: dbuf_size                            44,872   64,630,216   18,412,688
c: dbuf_size                            44,872   64,628,320    4,710,928
```

Maybe it needs a `dbuf_cache_min_bytes` and/or `dbuf_cache_min_pct` (as a function of ...`arc_meta_limit`?) to prevent the dbuf cache collapsing completely as a result of `arc_c` getting low. Or not.

**Comment by loli10K** [comment]
> Like this? chrisrd/zfs@9fa593d

Precisely, and looks like the "`[arc_reclaim]` burning a core"-thing i was seeing is not related to this change, it happens when i rapidly fill the ARC running `find` on a ~1mil 0-byte files even on https://github.com/zfsonlinux/zfs/commit/d3c2ae1c0806b183a315e3d43cc8018cfdca79b5.

I only have one question: why `arc_c`/`arc_target_bytes()` and not `arc_size`/`arc_bytes()`?

It would be nice if you could open a pull request, thanks.

**Comment by chrisrd** [comment]
The thinking for `arc_c` over `arc_size` was we want to help get where we're going (i.e. the target) rather than concentrate on where we are. Currently running local `zfs-tests` b/f opening pull request.

**Comment by dbavatar** [comment]
@chrisrd

> It looks like there's no mechanism to drop the dbuf cache below it's low water mark, e.g. in response to echo 3 > /proc/sys/vm/drop_caches. It's not clear if there should be or not.

I think we need it.

> I.e. it looks like the increase in metadata_size seen after d3c2ae1 is deliberate - mostly if not entirely due to the dbuf cache - rather than a leak per the topic of this ticket.

I don't think it's a leak in the traditional sense, but in the behavioral sense. You just can't get rid of the memory while running, and keeps growing for every file you access.

> Resolved?

Although the title of the issue has to do with metadata, the real issue here is we can't shrink zfs memory footprint to 0 (or minimum settings) via shrinker callback, and that opens up pandora's box of system level problems. Biggest one being kernel memory fragmentation failing high order allocations, long before you're actually out of memory. Also the memory footprint now is both high in absolute terms, and relatively to 6.5.x branch.
